// @xmtp/content-type-markdown/dist/index.js
import { ContentTypeId } from '@xmtp/content-type-primitives';

const ContentTypeMarkdown = new ContentTypeId({
    authorityId: "xmtp.org",
    typeId: "markdown",
    versionMajor: 1,
    versionMinor: 0,
});
var Encoding;
(function (Encoding) {
    Encoding["utf8"] = "UTF-8";
    Encoding["unknown"] = "unknown";
})(Encoding || (Encoding = {}));
class MarkdownCodec {
    get contentType() {
        return ContentTypeMarkdown;
    }
    encode(content) {
        return {
            type: ContentTypeMarkdown,
            parameters: { encoding: Encoding.utf8 },
            content: new TextEncoder().encode(content),
        };
    }
    decode(content) {
        if (content.parameters.encoding !== Encoding.utf8) {
            throw new Error(`unrecognized encoding ${content.parameters.encoding}`);
        }
        return new TextDecoder().decode(content.content);
    }
    fallback() {
        return undefined;
    }
    shouldPush() {
        return true;
    }
}

export { ContentTypeMarkdown, Encoding, MarkdownCodec };
