// @xmtp/agent-sdk/dist/core/ConversationContext.js
import { ContentTypeMarkdown } from "@xmtp/content-type-markdown";
import { ContentTypeText } from "@xmtp/content-type-text";
import { filter } from "../core/filter.js";
import { ClientContext } from "./ClientContext.js";
export class ConversationContext extends ClientContext {
    #conversation;
    constructor({ conversation, client, }) {
        super({ client });
        this.#conversation = conversation;
    }
    isDm() {
        return filter.isDM(this.#conversation);
    }
    isGroup() {
        return filter.isGroup(this.#conversation);
    }
    // Send methods, which don't need a message context, are in ConversationContext to make them available in both dm and group event handlers
    async sendMarkdown(markdown) {
        await this.conversation.send(markdown, ContentTypeMarkdown);
    }
    async sendText(text) {
        await this.conversation.send(text, ContentTypeText);
    }
    get conversation() {
        return this.#conversation;
    }
    get isAllowed() {
        return this.#conversation.consentState === 1 /* ConsentState.Allowed */;
    }
    get isDenied() {
        return this.#conversation.consentState === 2 /* ConsentState.Denied */;
    }
    get isUnknown() {
        return this.#conversation.consentState === 0 /* ConsentState.Unknown */;
    }
}
