// @xmtp/agent-sdk/dist/core/MessageContext.js
import { ContentTypeMarkdown, } from "@xmtp/content-type-markdown";
import { ContentTypeReaction, } from "@xmtp/content-type-reaction";
import { ContentTypeReply, } from "@xmtp/content-type-reply";
import { ContentTypeText } from "@xmtp/content-type-text";
import { filter } from "../core/filter.js";
import { ConversationContext } from "./ConversationContext.js";
export class MessageContext extends ConversationContext {
    #message;
    constructor({ message, conversation, client, }) {
        super({ conversation, client });
        this.#message = message;
    }
    usesCodec(codecClass) {
        return filter.usesCodec(this.#message, codecClass);
    }
    isMarkdown() {
        return filter.isMarkdown(this.#message);
    }
    isText() {
        return filter.isText(this.#message);
    }
    isReply() {
        return filter.isReply(this.#message);
    }
    isReaction() {
        return filter.isReaction(this.#message);
    }
    isReadReceipt() {
        return filter.isReadReceipt(this.#message);
    }
    isRemoteAttachment() {
        return filter.isRemoteAttachment(this.#message);
    }
    isTransactionReference() {
        return filter.isTransactionReference(this.#message);
    }
    isWalletSendCalls() {
        return filter.isWalletSendCalls(this.#message);
    }
    async sendReaction(content, schema = "unicode") {
        const reaction = {
            action: "added",
            reference: this.#message.id,
            referenceInboxId: this.#message.senderInboxId,
            schema,
            content,
        };
        await this.conversation.send(reaction, ContentTypeReaction);
    }
    async #sendReply(text, contentType = ContentTypeText) {
        const reply = {
            reference: this.#message.id,
            referenceInboxId: this.#message.senderInboxId,
            contentType,
            content: text,
        };
        await this.conversation.send(reply, ContentTypeReply);
    }
    async sendMarkdownReply(markdown) {
        await this.#sendReply(markdown, ContentTypeMarkdown);
    }
    async sendTextReply(text) {
        await this.#sendReply(text, ContentTypeText);
    }
    async getSenderAddress() {
        const inboxState = await this.client.preferences.inboxStateFromInboxIds([
            this.#message.senderInboxId,
        ]);
        return inboxState[0]?.identifiers[0]?.identifier;
    }
    get message() {
        return this.#message;
    }
}
